### PR TITLE
Feature/489169 extracting assessment of feasibility

### PR DESCRIPTION
**Pull request by otrnovec**
getting the assesment of feasibility for projects, some projects have none, some have multiple. Saved to .csv

### CODE REVIEW

**Comment by otrnovec** [comment]
only look at the extracting_assesment_of_feasibility.py for the purposes of this request

File: extracting_assesment_of_feasibility.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    # convert lists to dataframe
+    output_df = pd.DataFrame(
+        list_of_lists, columns=["ID", "Subjekt", "Odůvodnění", "Závěr"]
+    )
+
+    # save the dataframe to a csv file
+    output_df.to_csv("assesment_of_feasibility.csv", sep=",", encoding="utf-8")
+
+
+# #convert csv to df
```
Comment (PavelOujesky) [inline extracting_assesment_of_feasibility.py]: Two characters for a comment

**Review by PavelOujesky — APPROVED** [review]
The code seems fine to me, I just noticed that most comments don't have a space after the "#" character, which by convention they should. I don't fully understand some parts, but I believe it's fine.

File: extracting_assesment_of_feasibility.py (hunk 1 above)
Comment (otrnovec) [reply extracting_assesment_of_feasibility.py]: fixed

Comment (janrygl) [reply extracting_assesment_of_feasibility.py]: ... I don't fully understand some parts, but I believe it's fine.

@PavelOujesky That is not reassuring.:-) I recommend to put such code into LLMs with prompts like: I am learning to program. Please explain what does the following Python code do, line by line.

"""
code
"""

Comment (PavelOujesky) [reply extracting_assesment_of_feasibility.py]: Okay, I will do that.

File: extract_project_descriptions_and_public_interests.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        #extract the public interest from xml and save to the list
+        project_public_interest = soup.find("div", class_="col-xs-12 col-sm-offset-2 col-sm-8")
+        project_public_interest = project_public_interest.find("p")
+        project_public_interest = project_public_interest.find("i").text
+        #remove leading and trailing whitespaces
+        project_public_interest = project_public_interest.strip()
+        list_of_public_interest.append(project_public_interest)
+
+        #delay in seconds so we dont get blocked by the site
```
Comment (janrygl) [inline extract_project_descriptions_and_public_interests.py]: define DELAY in top of file, constants shouldn't be defined in functions

File: extract_project_descriptions_and_public_interests.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        #get the url content
+        url = requests.get(url, timeout=5)
+        soup = BeautifulSoup(url.content, 'html.parser')
+
+        #extract the project description from xml and save to the list
+        project_description = soup.find("div", class_="project-description").text
+        #remove the titles "Hlasovací video" and "Realizační video" from the description
+        project_description = re.sub('Hlasovací video', '', project_description)
+        project_description = re.sub('Realizační video', '', project_description)
```
Comment (janrygl) [inline extract_project_descriptions_and_public_interests.py]: make it oneliner
project_description = re.sub('(Hlasovací|Realizační) video', '', project_description)

File: extracting_assesment_of_feasibility.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            list_of_lists.append(assesment_list)
+
+    return list_of_lists
+
+
+def save_to_csv(list_of_lists):
+    """Function to save the assesments to a csv file."""
+    # convert lists to dataframe
+    output_df = pd.DataFrame(
+        list_of_lists, columns=["ID", "Subjekt", "Odůvodnění", "Závěr"]
```
Comment (janrygl) [inline extracting_assesment_of_feasibility.py]: I would suggest to use English name of columns

File: extracting_assesment_of_feasibility.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    list_of_lists = []
+
+    # loop through the urls in the csv
+    for url in url_df:
+        # get project_id from url and save to the list
+        project_id = url.split("id=", 1)[1]
+        assesment_list = [project_id]
+
+        # get the url content
+        url = requests.get(url, timeout=5)
```
Comment (janrygl) [inline extracting_assesment_of_feasibility.py]: put to top of file
DEFAULT_TIMEOUT = 5
and use it.
It will be easier to change default later

**Review by janrygl — APPROVED** [review]
Fix paths (windows vs linux slashes, data folder), less comments, but nice and clean code.

File: extract_project_descriptions_and_public_interests.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import re
+import requests
+from bs4 import BeautifulSoup
+import pandas as pd
+
+
+def extract_project_description_and_public_interest(url_df):
+    """Function to extract project description and public interest from the xml."""
+
+    #initialize lists
```
Comment (janrygl) [inline extract_project_descriptions_and_public_interests.py]: no comment needed

File: extract_project_descriptions_and_public_interests.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    list_of_project_ids = []
+    list_of_project_descriptions = []
+    list_of_public_interest = []
+
+    #loop through the urls in the csv
+    for url in url_df:
+        #get project_id from url and save to the list
+        project_id = url.split("id=",1)[1]
```
Comment (janrygl) [inline extract_project_descriptions_and_public_interests.py]: pep8, space after comma

File: extract_project_descriptions_and_public_interests.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
+""" This file is for getting the project descriptions and 
+public interests from xml and save them to a csv file. """
+import time
+import re
+import requests
+from bs4 import BeautifulSoup
+import pandas as pd
+
+
+def extract_project_description_and_public_interest(url_df):
```
Comment (janrygl) [inline extract_project_descriptions_and_public_interests.py]: I don't know type of url_df: is it a list of urls?
`def extract_project_description_and_public_interest(url_df: list[str]):`

File: extract_project_descriptions_and_public_interests.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        #get the url content
+        url = requests.get(url, timeout=5)
+        soup = BeautifulSoup(url.content, 'html.parser')
+
+        #extract the project description from xml and save to the list
+        project_description = soup.find("div", class_="project-description").text
+        #remove the titles "Hlasovací video" and "Realizační video" from the description
+        project_description = re.sub('Hlasovací video', '', project_description)
+        project_description = re.sub('Realizační video', '', project_description)
+        #remove leading and trailing whitespaces
```
Comment (janrygl) [inline extract_project_descriptions_and_public_interests.py:38]: comment describe strip function = no need

File: extract_project_descriptions_and_public_interests.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        #extract the public interest from xml and save to the list
+        project_public_interest = soup.find("div", class_="col-xs-12 col-sm-offset-2 col-sm-8")
+        project_public_interest = project_public_interest.find("p")
+        project_public_interest = project_public_interest.find("i").text
+        #remove leading and trailing whitespaces
+        project_public_interest = project_public_interest.strip()
+        list_of_public_interest.append(project_public_interest)
+
+        #delay in seconds so we dont get blocked by the site
+        DELAY = 1
```
Comment (janrygl) [inline extract_project_descriptions_and_public_interests.py]: DEFAULT_TIMEOUT here?

File: extracting_assesment_of_feasibility.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        else:
+            assesment_list.append("missing assesment of feasibility")
+            assesment_list.append("missing assesment of feasibility")
+            assesment_list.append("missing assesment of feasibility")
+            list_of_lists.append(assesment_list)
+
+    return list_of_lists
+
+
+def save_to_csv(list_of_lists):
```
Comment (janrygl) [inline extracting_assesment_of_feasibility.py]: still output path not in arg

File: extracting_assesment_of_feasibility.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+                count += 1
+                if count == 3:
+                    list_of_lists.append(assesment_list)
+                    assesment_list = [project_id]
+                    count = 0
+        # if the assesment is missing, save "missing assesment of feasibility"
+        else:
+            assesment_list.append("missing assesment of feasibility")
+            assesment_list.append("missing assesment of feasibility")
+            assesment_list.append("missing assesment of feasibility")
```
Comment (janrygl) [inline extracting_assesment_of_feasibility.py]: assesment_list.extend(["missing assesment of feasibility"] * 3)

**Review by janrygl — APPROVED** [review]

**Pull request merged**